core/model/src/main/java/org/eclipse/rdf4j/model/util/GraphComparisons.java

```diff
@@ -0,0 +1,415 @@
+				}
+				final Model canonicalizedContext1 = isoCanonicalize(contextInlModel1);
+				final Model canonicalizedContext2 = isoCanonicalize(contextInModel2);
+				if (!canonicalizedContext1.equals(canonicalizedContext2)) {
+					return false;
+				}
+			}
+			return true;
+		} else {
+			// only one context (the null context), so we're dealing with one graph only.
+			final Map<BNode, HashCode> mapping1 = getIsoCanonicalMapping(model1);
+			if (mapping1.isEmpty()) {
+				// no blank nodes in model1 - simple collection equality will do
+				return model1.equals(model2);
+			}
+			final Map<BNode, HashCode> mapping2 = getIsoCanonicalMapping(model2);
+			if (mappingsIncompatible(mapping1, mapping2)) {
+				return false;
+			}
+
+			final Model c1 = labelModel(model1, mapping1);
+			final Model c2 = labelModel(model2, mapping2);
+			return c1.equals(c2);
+		}
+	}
+
+	private static boolean mappingsIncompatible(Map<BNode, HashCode> mapping1, Map<BNode, HashCode> mapping2) {
+		if (mapping1.size() != mapping2.size()) {
+			return true;
+		}
+		Collection<HashCode> values1 = mapping1.values();
+		Collection<HashCode> values2 = mapping2.values();
+
+		if (!(values1.containsAll(values2) && values2.containsAll(values1))) {
+			return true;
+		}
+
+		return false;
+	}
+
+	protected static Model isoCanonicalize(Model m) {
+		return labelModel(m, getIsoCanonicalMapping(m));
+	}
+
+	protected static Map<BNode, HashCode> getIsoCanonicalMapping(Model m) {
+		Map<BNode, HashCode> blankNodeMapping = hashBNodes(m);
+		Multimap<HashCode, BNode> partition = partitionMapping(blankNodeMapping);
+
+		if (isFine(partition)) {
+			return blankNodeMapping;
+		}
+
+		return distinguish(m, blankNodeMapping, partition, null, new ArrayList<>(), new ArrayList<>());
+
+	}
+
+	protected static Set<BNode> getBlankNodes(Model m) {
+		final Set<BNode> blankNodes = new HashSet<>();
+
+		m.subjects().forEach(s -> {
+			if (s.isBNode()) {
+				blankNodes.add((BNode) s);
+			}
+		});
+		m.objects().forEach(o -> {
+			if (o.isBNode()) {
+				blankNodes.add((BNode) o);
+			}
+		});
+
+		return blankNodes;
+	}
+
+	private static Map<BNode, HashCode> distinguish(Model m, Map<BNode, HashCode> blankNodeMapping,
+			Multimap<HashCode, BNode> partitionMapping,
+			Map<BNode, HashCode> lowestFound, List<BNode> parentFixpoints,
+			List<Map<BNode, HashCode>> finePartitionMappings) {
+
+		final List<Collection<BNode>> sortedPartitions = new ArrayList<>(partitionMapping.asMap().values());
+		Collections.sort(sortedPartitions, new Comparator<Collection<BNode>>() {
+			public int compare(Collection<BNode> a, Collection<BNode> b) {
+				int result = a.size() - b.size();
+				if (result == 0) {
+					// break tie by comparing value hash
+					HashCode hashOfA = blankNodeMapping.get(a.iterator().next());
+					HashCode hashOfB = blankNodeMapping.get(b.iterator().next());
+					BigInteger difference = new BigInteger(1, hashOfA.asBytes())
+							.subtract(new BigInteger(1, hashOfB.asBytes()));
```

Comment:
If we used a 64bit hash I suppose we could `long` instead for added performance. What would be the trade offs?

Reply:
The main reason for using a 128 bit hash is that it gives us a hash that is, for all practical purposes, globally unique (so not just unique within the graph). This means we can use these values as unique identifiers for skolemization.

Reply:
Fwiw (see also below) I've switched it down to a 32 bit hash and using long instead of BigInteger. We can get away with this for isomorphism purposes (but not for skolemization). Oh and long instead of int (in case you're wondering) to handle potential overflows when we combine hashes.